### PR TITLE
Implement PCZT support

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/sapling-crypto/compare/01a5d1d20c600436be223cfd9334fa64b4ca696a..555bf581fd9731fde9f680627981a855da99c86d) to fix a bug and add missing serialization methods.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/sapling-crypto/compare/555bf581fd9731fde9f680627981a855da99c86d..290408a60ebae3e4e612c16646835082380a60df) to change `Output::rseed` to return `&[u8; 32]` instead of `Rseed` (and internally enforce that builders used to generate PCZTs require post-ZIP 212 outputs).

File: src/pczt/io_finalizer.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                    .sign(sighash, &ask, &mut rng)
+                    .map_err(IoFinalizerError::DummySignature)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Errors that can occur while extracting a regular Orchard bundle from a PCZT bundle.
```
Comment (str4d) [inline src/pczt/io_finalizer.rs:79]: Fix this comment.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/sapling-crypto/compare/290408a60ebae3e4e612c16646835082380a60df..7696219bf3740f6d979b2b7362e6c11154004010) to fix some bugs.

File: src/pczt.rs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    pub(crate) cmu: ExtractedNoteCommitment,
+
+    /// The ephemeral key used to encrypt the note plaintext.
+    pub(crate) ephemeral_key: EphemeralKeyBytes,
+
+    /// The encrypted note plaintext for the output.
+    ///
+    /// Once we have memo bundles, we will be able to set memos independently of Outputs.
+    /// For now, the Constructor sets both at the same time.
+    pub(crate) enc_ciphertext: [u8; ENC_CIPHERTEXT_SIZE],
```
Comment (nuttycom) [inline src/pczt.rs:203]: Should `Output` be polymorphic in the type here just so that the type doesn't have to change with ZIP 231?

**Review by nuttycom — APPROVED** [review]
utACK 7696219bf3740f6d979b2b7362e6c11154004010 with nonblocking public API future compatibility question.

File: src/pczt.rs (hunk 2 above)
Comment (str4d) [reply src/pczt.rs:203]: I don't have the time to design that such that I can ensure it's forward compatible. And we'll need to change much of the rest of the API anyway, so this isn't much extra.

File: src/builder.rs
Comment (nuttycom) [inline src/builder.rs:1]: I'm running into a problem using the builder; `add_spend` still expects the `ExtendedSpendingKey` to be provided, which won't work for our desired Keystone use case.

File: src/builder.rs [hunk 3]
```diff
@@ -310,11 +330,11 @@ impl PreparedSpendInfo {
             value: Some(self.note.value()),
             rseed: Some(*self.note.rseed()),
             rcv: Some(self.rcv),
-            proof_generation_key: Some(self.proof_generation_key),
+            proof_generation_key: None,
```
[the remark this answers is not in the document]
Comment (str4d) [reply src/builder.rs:333]: This was actually a bug: while for regular spends we want this to be `None`, for dummy spends we still need to set the proof generation key here, or else we become unable to do so later. I was correctly handling this case for regular builds, but missed it here.

File: src/builder.rs [hunk 4]
```diff
@@ -207,48 +222,78 @@ impl SpendInfo {
...
+
+    fn build<Pr: SpendProver, R: RngCore>(
+        self,
+        proof_generation_key: Option<ProofGenerationKey>,
+        rng: R,
+    ) -> Result<SpendDescription<InProgress<Unproven, Unsigned>>, Error> {
+        let proof_generation_key = match (proof_generation_key, self.dummy_expsk.as_ref()) {
+            (Some(proof_generation_key), None) => Ok(proof_generation_key),
+            (None, Some(expsk)) => Ok(expsk.proof_generation_key()),
+            (Some(_), Some(_)) => Err(Error::WrongSpendingKey),
```
Comment (daira) [inline src/builder.rs:279]: Should the case where `proof_generation_key` is `Some` and equal to `expsk.proof_generation_key()` be permitted?
```suggestion
            (Some(proof_generation_key), Some(expsk)) if proof_generation_key == expsk.proof_generation_key() => Ok(proof_generation_key),
            (Some(_), Some(_)) => Err(Error::WrongSpendingKey),
```

File: src/builder.rs [hunk 5]
```diff
@@ -547,20 +675,150 @@ impl Builder {
...
     }
+
+    /// Builds a bundle containing the given spent notes and outputs along with their
+    /// metadata, for inclusion in a PCZT.
+    pub fn build_for_pczt(
+        self,
+        rng: impl RngCore,
+    ) -> Result<(crate::pczt::Bundle, SaplingMetadata), Error> {
+        match self.zip212_enforcement {
+            Zip212Enforcement::Off | Zip212Enforcement::GracePeriod => {
```
Comment (daira) [inline src/builder.rs:689]: Why fail in the grace period? (Non-blocking; only matters in practice for tests.)

File: src/pczt.rs (hunk 2 above)
Comment (daira) [reply src/pczt.rs:206]: For Orchard it's a `Vec<u8>`. Isn't it better to do the same here for consistency?

File: src/pczt/signer.rs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        ask: &SpendAuthorizingKey,
+        rng: R,
+    ) -> Result<(), SignerError> {
+        let alpha = self.alpha.ok_or(SignerError::MissingSpendAuthRandomizer)?;
+
+        let rsk = ask.randomize(&alpha);
+        let rk = redjubjub::VerificationKey::from(&rsk);
+
+        if self.rk == rk {
+            self.spend_auth_sig = Some(rsk.sign(rng, &sighash));
```
Comment (daira) [inline src/pczt/signer.rs:23]: I went down a rabbit hole here verifying that https://github.com/zcash/librustzcash/issues/179 is fixed. [It is.](https://github.com/zcash/librustzcash/issues/179#issuecomment-2537459745)

**Review by daira — APPROVED** [review]
utACK with non-blocking comments.

File: src/builder.rs (hunk 4 above)
Comment (str4d) [reply src/builder.rs:279]: `self.dummy_expsk` is generated internally from the RNG, and isn't directly extractable.

File: src/pczt.rs (hunk 2 above)
Comment (str4d) [reply src/pczt.rs:206]: [It already does](https://github.com/zcash/librustzcash/blob/205d4c930319b7b6d24aeb4efde69e9b4d1b6f7b/pczt/src/sapling.rs#L165-L177).

**Pull request merged**